src/coap.c

```diff
@@ -0,0 +1,311 @@
+   * reserved, MUST NOT be sent, and MUST be processed as a message
+   * format error.
+   */
+  cm->token.len = *ptr & 0x0F;
+  if (cm->token.len > 8) {
+    cm->flags |= COAP_FORMAT_ERROR;
+    return NULL;
+  }
+
+  ptr++;
+
+  /*
+   * Code:  8-bit unsigned integer, split into a 3-bit class (most
+   * significant bits) and a 5-bit detail (least significant bits)
+   */
+  cm->code_class = *ptr >> 5;
+  cm->code_detail = *ptr & 0x1F;
+  cm->flags |= (COAP_CODE_CLASS_FIELD | COAP_CODE_DETAIL_FIELD);
+
+  ptr++;
+
+  /* Message ID:  16-bit unsigned integer in network byte order. */
+  cm->msg_id = (uint8_t)*ptr << 8 | (uint8_t)*(ptr + 1);
+  cm->flags |= COAP_MSG_ID_FIELD;
+
+  ptr += 2;
+
+  return ptr;
+}
+
+/* Fulls token information in cm */
+static char *coap_get_token(char *ptr, struct iobuf *io,
+                            struct ns_coap_message *cm) {
+  if (cm->token.len != 0) {
+    if (ptr + cm->token.len > io->buf + io->len ) {
+      cm->flags |= COAP_NOT_ENOUGH_DATA;
+      return NULL;
+    } else {
+      cm->token.p = ptr;
+      ptr += cm->token.len;
+      cm->flags |= COAP_TOKEN_FIELD;
+    }
+  }
+
+  return ptr;
+}
+
+/* Returns Option Delta or Length. Helper function. */
+static int coap_get_ext_opt(char* ptr, struct iobuf *io, uint16_t* opt_info) {
+  int ret = 0;
+
+  if (*opt_info == 13) {
+    /*
+     * 13:  An 8-bit unsigned integer follows the initial byte and
+     * indicates the Option Delta/Length minus 13.
+     */
+    if (ptr < io->buf + io->len) {
+      *opt_info = *ptr + 13;
+      ret = sizeof(uint8_t);
+    } else {
+      ret = -1;
+    }
+  } else if (*opt_info == 14) {
+    /*
+     * 14:  A 16-bit unsigned integer in network byte order follows the
+     * initial byte and indicates the Option Delta/Length minus 269.
+     */
+    if (ptr + sizeof(uint8_t) < io->buf + io->len) {
+      *opt_info = (uint8_t)*ptr << 8 | (uint8_t)*(ptr + 1) + 269;
+      ret = sizeof(uint16_t);
+    } else {
+      ret = -1;
+    }
+  }
+
+  return ret;
+}
+
+/* 
+ * Fills options in cm.
+ *
+ * General options format:
+ * +---------------+---------------+
+ * | Option Delta  | Option Length |  1 byte
+ * +---------------+---------------+
+ * \    Option Delta (extended)    \  0-2 bytes
+ * +-------------------------------+
+ * / Option Length  (extended)     \  0-2 bytes
+ * +-------------------------------+
+ * \         Option Value          \  0 or more bytes
+ * +-------------------------------+
+ */
+static char *coap_get_options(char* ptr, struct iobuf *io,
+                              struct ns_coap_message *cm) {
+  uint16_t prev_opt = 0;
+
+  if (ptr == io->buf + io->len) {
+    /* end of packet, ok */
+    return NULL;
+  }
+
+  /* 0xFF is payload marker */
+  while ((unsigned char)*ptr != 0xFF && ptr < io->buf + io->len) {
+    uint16_t option_delta, option_lenght;
+    int optinfo_len;
+
+    /* Option Delta:  4-bit unsigned integer */
+    option_delta = (*ptr & 0xF0) >> 4;
```

Comment:
note: this works because `*ptr & 0xf0` yields an unsigned expression because the `0xf0` liter, which then get shifted to the right without sign extension. 

This depends on the fact that the compiler treats `0xF0` as an unsigned value. In fact `(*ptr & (char)0xF0) >> 4` would yield 255.

I'm not sure if this behaviour is well defined by the C standard.